### DISCUSSION TITLE
Why makeAutoObservable does not support subclassing

### DISCUSSION THREAD

**Discussion by cncolder** [mobxjs/mobx#2850] [Q&A] [answered]
I have read the document https://mobx.js.org/subclassing.html, and I understand the `makeAutoObservable` does not support subclassing now.

But I'm not sure the reason. In fact, I try to comment out this line https://github.com/mobxjs/mobx/blob/0945c26513057457e1534a80558a3eb98487dc96/packages/mobx/src/api/makeObservable.ts#L52. It works as expects. 🤔 

```
class BaseStore {
  data = {}
}

class Store extends BaseStore {
  constructor() {
    super();
    makeAutoObservable(this);
  }
  
  loading = false
}
```

**Comment by urugator** [comment, marked as answer]
The situation where the extended classes are abstract (can't be instantiated) - meaning you can call `makeAutoObservable` only once in the "subbest" class - is kinda supportable, the problem is if the extended class is eg `React.Component` - in such case `makeAutoObservable` should ignore all it's fields and prototype.
In other cases, you must be able to call `makeAutoObservable` multiple times on the same object - this is problematic, because we don't know which fields of the object were already annotated and which were not (this is mostly due to the ability to provide `overrides` option to  `makeAutoObservable` or when you would combine it with `makeObservable`). It would also prevented some optimizations.
We don't have this issue with `makeObservable`, because user is the one providing the list of fields to annotate.
Note, there is no way of telling, which field belongs to which class - eg in your example we don't know that `data` field comes from `BaseClass`, similary we don't know from which constructor `make(Auto)Observable` is being called, therefore we had to impose some restrictions on subclassing and annotating. [More on that here.](https://github.com/mobxjs/mobx/pull/2641#issue-526745810)

**Reply by stephenh** [reply]
@urugator really appreciate your replies here, and overall work on mobx, really love it...

Per your disclaimers about being "kinda supportable ... if you follow these restrictions", I tried to put together an extremely limited version of `makeAutoObservable`, and AFAICT it's working for us:

https://gist.github.com/stephenh/77f62941913203a871d0e284ea779fe9

We have ~10 classes that inherit from a base class, but none of them do overriding of parent functionality (we're really using the base class more like a mixin).

I'd be curious, skimming that gist, does anything stand out about "oh you've forgotten about x/y/z that is totally going to break it?"

Thanks!

**Reply by urugator** [reply]
Didn't try to run, but something like following should work.
The gist is to collect all keys and just delegate to `makeObservable`, it will already handle all supers (as mentioned we don't know from where the `makeObservable` is called, so we always annotate whole proto chain)
```javascript
import * as mobx from "mobx";

const annotationsSymbol = new Symbol();
const objectPrototype = Object.prototype;

export function makeSimpleAutoObservable(target, overrides, options) {
  // Make sure nobody called makeObservable/makeAutoObservable/extendObservable/makeSimpleAutoObservable previously (eg in parent constructor)
  if (mobx.isObservable(target)) {
    throw new Error("Target must not be observable");
  }

  let annotations = target[annotationsSymbol];
  if (!annotations) {    	
    annotations = {}
    const proto = Object.getPrototypeOf(target)
    let current = target;
    while (current && current !== objectPrototype) {
      Reflect.ownKeys(current).forEach(key => {
        if (key === mobx.$mobx || key === 'constructor') return;
        annotations[key] = !overrides ? true : key in overrides ? overrides[key] : true
      });
      current = Object.getPrototypeOf(current);
    }
    // Cache if class
    if (proto && proto !== objectPrototype) {
      Object.defineProperty(proto, annotationsSymbol, { value: annotations });
    }
  }

  return mobx.makeObservable(target, annotations, options);
}
```
> Subclasses should not override parent class methods (although this might? work)

Atm `override` annotation is optional, so it should work, but I can't quarantee it won't ever change.

**Reply by LandonSchropp** [reply]
@stephenh This is awesome! Have you considered packaging this into a library?

**Reply by inoyakaigor** [reply]
### ⚠Notification for those who will use this code

This code does not check the equality that all proprties in the `annotations` exists in the `target`. If it does not Mobx will throw error [No. 1](/mobxjs/mobx/blob/main/packages/mobx/src/errors.ts#L4). You must manually add a check to make sure all properties exists.

UPD 2023: This fix was made in this package [mobx-store-inheritance](https://www.npmjs.com/package/mobx-store-inheritance)

**Reply by KashubaK** [reply]
@stephenh Is there a reason why `let current = proto;` shouldn't be `let current = target;`? In my usage it seemed to skip some of the properties that do belong to the `target` that resulted in a lack of reactivity. Using the latter assignment filled in the blanks and generated the annotations as I expected.

Also, nearly 2 years later, has there been any drawbacks or issues with this solution? Are you still using it, or did you find a more desirable solution?

**Reply by urugator** [reply]
> Is there a reason why let current = proto; shouldn't be let current = target;

It's probably wrong. As I said, I've never try to run the code above, it's only to demonstrate the idea. Do not copy-paste it to production code as is. I've changed the mentioned bit however.

**Reply by stephenh** [reply]
Ah yeah, @KashubaK you are right; our local `makeSimpleAutoObservable` had your suggested `current = target` fix as well, just as @urugator 's edit now does.

So we must have also fixed it back in the day, but apologies didn't think to post it back up. :-/

Thanks for updating @urugator !

**Reply by Kashuab** [reply]
Thanks guys for the responses :) Cheers!

**Reply by inoyakaigor** [reply]
The same story as @stephenh's. Made that fix local but forget mention it here

**Reply by inoyakaigor** [reply]
@LandonSchropp After a few years testing these code in production I finally made a package: [mobx-store-inheritance](https://www.npmjs.com/package/mobx-store-inheritance)

Folks FYI @stephenh @KashubaK @cncolder

**Reply by SnowdenWintermute** [reply]
Thank you! The main reason I switched to MobX is because it is the only state management solution that supports traditional OOP patterns. In my use case my objects are not "stores" in the traditional sense, but game entities that live both on a server and client that share common code. It is incredibly useful to be able to do something like this:

```
export class CombatantSubsystem {
  protected combatantProperties: CombatantProperties | undefined;

  initialize(combatantProperties: CombatantProperties) {
    this.combatantProperties = combatantProperties;
  }

  protected getCombatantProperties() {
    if (this.combatantProperties === undefined) {
      throw new Error(ERROR_MESSAGES.CLASS_INSTANCE_NOT_INITIALIZED);
    }
    return this.combatantProperties;
  }
}

class MySubsystem extends CombatantSubsystem{}
```

Otherwise we have to choose between maintaining a list of explicit observables (error prone and wordy) or copying initialization code in every subsystem (not maintainable when initialization code changes).

I wonder if there could be something like makeAutoObservableOnLeaf() which could only be called if all parent classes in the inheritance tree called something like allowAutoObservablesOnExtended() in their constructors?

I like MobX because it supports a lot of OOP techniques, and inheritance is a very powerful OOP technique.

**Comment by stephenh** [comment]
That's awesome! Thank you!

Fwiw this seems like a common ask from users, i.e. I nearly opened a brand
new discussion/issue asking the same question (just to ask specifically
what makeAutoObservables sharp edges are with subclasses to verify whether
we could avoid them and afaict we are/can).

What do you think about addong your snippet to mobx with an appropriate
name and docs of "only use this if you know what you are doing in these x
scenarios"?

Or if anything maybe a link in makeAutoObservables docs that "if you really
want subclasses see this gist that you can copy/paste and be on your own".
:-)

**Reply by urugator** [reply]
No, because this is not a suggested way to go, therefore I don't want to promote it in any way. Whatever you expose, some people will use it, twist it, misuse it, and in the end it will always come back to you, no matter the warnings.
Automatic behavior provides convinience for common/simple cases. If it can't support inheritance reliably, the responsbile thing to do is either not introducing automatic behavior in the first place (no longer an option) or to forbid cases that could lead to issues.

I am quite convinced that the cost of using `makeObservable` instead of `makeAutoObservable` (or even better getting rid of inheritance) for these ~10 classes , will be lower than the cost of introducing and maintaning any workaround.

**Reply by stephenh** [reply]
That's a depressing conclusion :-) but I understand its born from your real world experience, so I understand. Thanks!

**Reply by LandonSchropp** [reply]
@urugator Just to throw my 2¢ into the discussion, I'm not sure if I agree with your assessment. The problem is it's really easy to add properties and methods to your store and to forget to add them to `makeObservable`. This can lead to nasty bugs where components updates when another property's value changes, but doesn't if the missing property is changed. These types of issues can be very tricky to track down. Personally, I'd rather have things handled automatically.

**Reply by urugator** [reply]
> This can lead to nasty bugs

Ok but as mentioned the alternative is a solution that:
- can lead to other bugs (when extending foreign classes, which is impossible to detect)
- supports only extending abstract classes (again impossible to detect)
- requires different (and weird) usage than rest of the API (`makeObservable` should be called in each constructor, while `makeAutoObservable` only in the last constructor)

Unless you're dealing with existing codebase and moving away from decorators, it's not like you're forced to use inheritence and therefore use `makeObservable` and therefore introduce nasty bugs. You can use composition or you can use decorators, or you don't have to use classes... There are different approches with different (hopefully reasonable) tradeoffs, you just have to desing the system with these in mind and pick the one providing best value for your project.

That being said I think it could be interesting to have an eslint plugin (because it can't be done runtime), which checks that `makeObservable` lists all fields defined on a class.

EDIT: For future readers, we do have the plugin now https://github.com/mobxjs/mobx/tree/main/packages/eslint-plugin-mobx

**Comment by ThaJay** [comment]
I'm getting the message:
`[Error: [MobX] Extending an object with another observable (object) is not supported]`
without extra information.

I'm doing maintenance on an app that had development frozen for 2 years, so it's probably a migration issue with new package versions. Unfortunately, there is no stack trace or other hint what / where exactly it goes wrong, this would be nice to have.

I do have some subclassing, but that's only a common constructor to fill the stores with data. It used to work. I'll grab the docs and see if I can solve it. If I encounter something weird I'll post it here. For now, thanks for being a rubber duck :)

**Reply by urugator** [reply]
You're probably passing an observable object as second arg to `extendObservable` somewhere. It expects plain object instead.

**Comment by Lars-jnk** [comment]
Are there any plans to work on that issue?

**Reply by mweststrate** [reply]
no

**Reply by Lars-jnk** [reply]
That's not good.
Can you recommend an alternative that support inheritance?

**Reply by inoyakaigor** [reply]
Urugator wrote a sample of code that adds inheritance. I made from this code the [package](https://www.npmjs.com/package/mobx-store-inheritance)

**Comment by mweststrate** [comment]
MakeObservable supports inheritance, just  the auto version doesn't

**Reply by Lars-jnk** [reply]
Thanks for the hint. I will look into that.

**Comment by tabekg** [comment]
if I have property #createdAt, and there are getter createdAt, setter createdAt. How can I make observable this getter and setter? @mweststrate

**Comment by js2me** [comment]
I'll just leave this here, I think it will be useful.  

https://www.npmjs.com/package/mobx-store-inheritance  

Author: @inoyakaigor

**Reply by inoyakaigor** [reply]
I took the code for this package right from this discussion 😄. Have a look at the other threads :)
